### ISSUE TITLE
[Configuration cache] Dependent OsDetectorPlugin is not compliant with configuration cache

### ISSUE THREAD

**Issue by voidzcy** [google/protobuf-gradle-plugin#409]
As mentioned in https://github.com/google/protobuf-gradle-plugin/pull/406#discussion_r434116323, the dependent OsDetectorPlugin uses JDK APIs to read system properties, which violates the [requirement for configuration cache](https://docs.gradle.org/nightly/userguide/configuration_cache.html#requirements):

> Plugins and build scripts should not read system properties directly using the Java APIs at configuration time. Instead, these system properties must be declared as a potential build input by using the value supplier APIs.

That needs to be fixed either in osdetector-gradle-plugin or having some workaround in protobuf-gradle-plugin if possible.

Users of protobuf-gradle-plugin are still able to take advantage of configuration cache feature with `--configuration-cache=warn` argument (ignoring errors of `InvalidUserCodeException`, which are considered to be harmless, our [existing test](https://github.com/google/protobuf-gradle-plugin/blob/a7b0c53185dff590e4d1850962bb97f29dc22dfa/src/test/groovy/com/google/protobuf/gradle/ProtobufJavaPluginTest.groovy#L94) does so).

**Comment by voidzcy** [comment]
There is no easy way to mitigate this now. 

- The osdetector-gradle-plugin delegates reading system properties to the os-maven-plugin, which calls JDK APIs.
- Reading system properties happens in configuration phase (the afterEvaluate hook), we are not able to wrap system property reading into a build service that is used in task execution.

We probably would need to migrate away from os-mavan-plugin.

**Comment by gavra0** [comment]
@voidzcy Can we implement this in the Protobuf plugin? We probably need to support just a couple of basic os/arch combinations.

**Comment by voidzcy** [comment]
We need to read the system property in configuration time to determine the dependency notation in a map form for protoc. I am not sure if a map with lazily evaluated values would work. 

Even with that, we would need to implement the system properties/files reading with Gradle's `Provider` approach by ourselves just in order to get the classifier to be used the dependency. Are you able to contribute?

**Comment by ejona86** [comment]
I don't see any need to reimplement this logic. It seems very easy to enhance os-detector-maven to work with our needs. It seems Detector is pretty close already. It already requires overriding some protected methods for logging; we would just add a few for accessing `System`. It looks like there is only one static method using `System`, and that can just be overloaded.

**Comment by voidzcy** [comment]
We can try wrapping `project.osdetector.classifier` into a Gradle `Provider` and chain it with [`Provider.forUseAtConfigurationTime()`](https://docs.gradle.org/nightly/javadoc/org/gradle/api/provider/Provider.html#forUseAtConfigurationTime--). A normal `Provider` would not work as it will just trigger the read operation to happen. But still my biggest concern is if the dependency notation can contain some lazily evaluated value and it cannot be resolved until the execution phase.

**Comment by voidzcy** [comment]
I tried using [`MapProperty`](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/MapProperty.html), but no luck. Would need more involved changes.

**Comment by voidzcy** [comment]
I've experimented locally by replacing
 
- `System.getProperty()` with `ProviderFactory.systemProperty(name).forUseAtConfigurationTime().getOrNull()`

- File reading with 
```java
RegularFile file = getProjectLayout().getProjectDirectory().file(fileName);
ProviderFactory.fileContents(file).getAsBytes().forUseAtConfigurationTime().getOrNull();
```

It works with configuration cache.

Note this approach still reads system properties/files at configuration phase (in a configuration cache compliant way), which means changing to their values will cause cache entry invalidation and reconfiguration. But since the os-detector plugin is mostly reading OS properties/files, they are not expect to change between builds in normal cases. So configuration cache would still be beneficial for most users.

I've created https://github.com/trustin/os-maven-plugin/pull/47 for os-maven-plugin. Once that gets accepted, I will make changes to osdetetor-gradle-plugin and then solve this issue here.
